File: app/src/services/websocket.ts
/**
 * WebSocket 服务 - 同步进度实时推送
 */

export type ProgressEventType = 
  | 'connected'
  | 'fetch_start'
  | 'fetch_progress'
  | 'fetch_complete'
  | 'save_start'
  | 'save_progress'
  | 'save_complete'
  | 'error'
  | 'info'
  | 'pong';

export interface ProgressEvent {
  type: ProgressEventType;
  message: string;
  current?: number;
  total?: number;
  timestamp?: string;
  sync_id?: string;
  inserted?: number;
  updated?: number;
}

type ProgressCallback = (event: ProgressEvent) => void;

class WebSocketService {
  private ws: WebSocket | null = null;
  private callbacks: Set<ProgressCallback> = new Set();
  private reconnectAttempts = 0;
  private maxReconnectAttempts = 5;
  private reconnectDelay = 3000;
  private url: string;

  constructor() {
    // 根据当前环境确定 WebSocket URL
    const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
    const host = window.location.host;
    this.url = `${protocol}//${host}/ws/sync-progress`;
  }

  /**
   * 连接 WebSocket
   */
  connect(): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      console.log('[WebSocket] 已经连接');
      return;
    }

    console.log('[WebSocket] 正在连接...', this.url);
    
    try {
      this.ws = new WebSocket(this.url);

      this.ws.onopen = () => {
        console.log('[WebSocket] 连接成功');
        this.reconnectAttempts = 0;
      };

      this.ws.onmessage = (event) => {
        try {
          const data: ProgressEvent = JSON.parse(event.data);
          console.log('[WebSocket] 收到消息:', data);
          this.notifyCallbacks(data);
        } catch (error) {
          console.error('[WebSocket] 消息解析失败:', error);
        }
      };

      this.ws.onerror = (error) => {
        console.error('[WebSocket] 连接错误:', error);
      };

      this.ws.onclose = () => {
        console.log('[WebSocket] 连接关闭');
        this.ws = null;
        this.attemptReconnect();
      };
    } catch (error) {
      console.error('[WebSocket] 连接失败:', error);
      this.attemptReconnect();
    }
  }

  /**
   * 尝试重新连接
   */
  private attemptReconnect(): void {
    if (this.reconnectAttempts >= this.maxReconnectAttempts) {
      console.log('[WebSocket] 达到最大重连次数，停止重连');
      return;
    }

    this.reconnectAttempts++;
    console.log(`[WebSocket] ${this.reconnectDelay}ms 后尝试第 ${this.reconnectAttempts} 次重连...`);
    
    setTimeout(() => {
      this.connect();
    }, this.reconnectDelay);
  }

  /**
   * 断开连接
   */
  disconnect(): void {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }

  /**
   * 注册进度回调
   */
  onProgress(callback: ProgressCallback): () => void {
    this.callbacks.add(callback);
    return () => {
      this.callbacks.delete(callback);
    };
  }

  /**
   * 通知所有回调
   */
  private notifyCallbacks(event: ProgressEvent): void {
    this.callbacks.forEach(callback => {
      try {
        callback(event);
      } catch (error) {
        console.error('[WebSocket] 回调执行失败:', error);
      }
    });
  }

  /**
   * 发送 ping
   */
  ping(): void {
    if (this.ws?.readyState === WebSocket.OPEN) {
      this.ws.send('ping');
    }
  }

  /**
   * 检查连接状态
   */
  isConnected(): boolean {
    return this.ws?.readyState === WebSocket.OPEN;
  }
}

// 单例实例
export const wsService = new WebSocketService();

File: backend/api/websocket.py
"""
WebSocket API - 同步进度实时推送
"""
from fastapi import APIRouter, WebSocket, WebSocketDisconnect
from typing import Set
from utils.progress import progress_manager

router = APIRouter(prefix="/ws", tags=["WebSocket"])


@router.websocket("/sync-progress")
async def sync_progress_websocket(websocket: WebSocket):
    """
    WebSocket 端点：同步进度实时推送
    
    连接后，所有同步器的进度更新都会通过此 WebSocket 推送
    
    消息格式：
    {
        "type": "fetch_start|fetch_progress|fetch_complete|save_start|save_progress|save_complete|error|info",
        "message": "进度描述",
        "current": 当前进度,
        "total": 总进度,
        "timestamp": "ISO格式时间戳",
        "sync_id": "同步任务ID",
        ...其他字段
    }
    """
    await websocket.accept()
    client_info = f"{websocket.client.host}:{websocket.client.port}"
    print(f"[WebSocket] 客户端连接: {client_info}")
    
    # 注册 WebSocket 连接
    progress_manager.register_websocket(websocket)
    
    try:
        # 发送连接成功消息
        await websocket.send_json({
            "type": "connected",
            "message": "已连接到同步进度推送服务",
            "timestamp": ""
        })
        
        # 保持连接，等待客户端消息（可选，用于心跳检测等）
        while True:
            try:
                # 接收客户端消息（非阻塞，超时 30 秒）
                data = await websocket.receive_text()
                # 可以处理客户端发送的命令，如 ping/pong
                if data == "ping":
                    await websocket.send_json({"type": "pong", "message": "ok"})
            except Exception:
                # 接收超时或出错，继续循环
                pass
                
    except WebSocketDisconnect:
        print(f"[WebSocket] 客户端断开连接: {client_info}")
    except Exception as e:
        print(f"[WebSocket] 连接异常: {client_info}, 错误: {e}")
    finally:
        # 注销 WebSocket 连接
        progress_manager.unregister_websocket(websocket)
        print(f"[WebSocket] 客户端已清理: {client_info}")
